modules/src/applications/transfer/relay/on_recv_packet.rs

```diff
@@ -0,0 +1,86 @@
+use crate::applications::transfer::error::Error as Ics20Error;
+use crate::applications::transfer::events::DenomTraceEvent;
+use crate::applications::transfer::packet::PacketData;
+use crate::applications::transfer::{IbcCoin, Source, TracePrefix};
+use crate::core::ics04_channel::packet::Packet;
+use crate::core::ics26_routing::context::{ModuleOutputBuilder, WriteFn};
+use crate::prelude::*;
+
+pub fn process_recv_packet<Ctx: 'static + Ics20Context>(
+    ctx: &Ctx,
+    output: &mut ModuleOutputBuilder,
+    packet: &Packet,
+    data: PacketData,
+) -> Result<Box<WriteFn>, Ics20Error> {
+    if !ctx.is_receive_enabled() {
+        return Err(Ics20Error::receive_disabled());
+    }
+
+    let receiver = data
+        .receiver
+        .clone()
+        .try_into()
+        .map_err(|_| Ics20Error::parse_account_failure())?;
+
+    let prefix = TracePrefix::new(packet.source_port.clone(), packet.source_channel);
+    match data.token.denom.source_chain(&prefix) {
+        Source::Receiver => {
+            // sender chain is not the source, unescrow tokens
+            let coin = {
+                let mut c = data.token.clone();
+                c.denom.remove_prefix(&prefix);
+                c
+            };
+
+            if ctx.is_blocked_account(&receiver) {
+                return Err(Ics20Error::unauthorised_receive(data.receiver));
+            }
+
+            let escrow_address = ctx
+                .get_channel_escrow_address(&packet.destination_port, packet.destination_channel)?;
+            let amount = IbcCoin::from(coin);
+
+            Ok(Box::new(move |ctx| {
+                let ctx = ctx.downcast_mut::<Ctx>().unwrap();
+                ctx.send_coins(&escrow_address, &receiver, &amount)
+                    .map_err(|e| e.to_string())
+            }))
+        }
+        Source::Sender => {
+            // sender chain is the source, mint vouchers
+            let prefix =
+                TracePrefix::new(packet.destination_port.clone(), packet.destination_channel);
+            let coin = {
+                let mut c = data.token;
+                c.denom.add_prefix(prefix);
+                c
+            };
+
+            let denom_trace_event = DenomTraceEvent {
+                trace_hash: coin.denom.hashed(),
+                denom: coin.denom.clone(),
+            };
+            output.emit(denom_trace_event.into());
+
+            Ok(Box::new(move |ctx| {
+                let ctx = ctx.downcast_mut::<Ctx>().unwrap();
+                let hashed_denom = coin.denom.hashed();
+                if ctx.has_denom_trace(&hashed_denom) {
+                    ctx.set_denom_trace(&coin.denom)
+                        .map_err(|e| e.to_string())?;
+                }
+
+                let amount = IbcCoin::from(coin);
+                ctx.mint_coins(&ctx.get_transfer_account(), &amount)
+                    .map_err(|e| e.to_string())?;
+                ctx.send_coins_from_module_to_account(
```

Comment:
IIUC, the SDK only allows module accounts to mint coins to restrict the ability to mint coins to certain 'privileged' accounts. On our side, both module accounts and user accounts are identified using `AccountId`s, so I think it makes sense to have different functions to send coins from user accounts to other user accounts and module accounts to user accounts. We can change the naming of functions to not use 'module' accounts and use privileged accounts or something or we can also use a different type to represent these accounts instead of `AccountId`.

Reply:
But that's an implementation detail for the SDK, no? Our trait could ask the chain "mint coins to this user address", and the chain can do so however it wishes. In the case of the SDK, it would mint coins at the `transfer` module address, and send the coins from there to the user account (as we do here).

The chain can use whatever security policy for minting coins as it wishes. If it doesn't like the address that we request the `mint` to, then it can fail the request. We don't need to know why and how it got to that conclusion.

In other words, I could currently add a method `mint_coins_to_user_addr()` to the trait, and implement as we do here (mint to module account, then send to user addr), without changing the SDK's policy that coins can only be minted at module addresses.

Reply:
You're right 👍 -> https://github.com/informalsystems/ibc-rs/pull/1989/commits/a75453db9074da904655b3142caaaca37bb57a41